src/metabase/driver/h2.clj

```diff
@@ -127,22 +133,51 @@
+
+  - Each `command-type` corresponds to a value in org.h2.command.CommandInterface, and match the commands in order.
+  - `remaining-sql` is a nillable sql string that is unable to be classified without running preceding queries first.
+    Usually if `remaining-sql` exists we will deny the query."
   [database query]
   (when-let [h2-parser (make-h2-parser database)]
     (try
-      (let [command      (.prepareCommand h2-parser query)
-            command-type (.getCommandType command)]
-        ;; TODO: do we need to handle CommandList?
-        ;; Command types are organized with all DDL commands listed first
-        ;; see https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java
-        (< command-type CommandInterface/ALTER_SEQUENCE))
-      ;; if the query is invalid, then it isn't DDL
-      (catch Throwable _ false))))
-
-(defn- check-disallow-ddl-commands [{:keys [database] {:keys [query]} :native}]
-  (when (and query (contains-ddl? database query))
-    (throw (IllegalArgumentException. "DDL commands are not allowed to be used with h2."))))
+      (let [command-list      (.prepareCommand h2-parser query)
+            command-list-type (.getCommandType ^org.h2.command.CommandList command-list)
+            command-types     (cond-> [command-list-type]
+                                (not (instance? org.h2.command.CommandContainer command-list))
+                                (into
+                                 (map #(.getType ^org.h2.command.Prepared %))
+                                 ;; when there are no fields: return no commands
+                                 (get-field command-list "commands" [])))]
+        {:command-types command-types
+         ;; when there is no remaining sql: return nil for remaining-sql
+         :remaining-sql (get-field command-list "remaining" nil)})
+      ;; only valid queries can be classified.
+      (catch org.h2.message.DbException _
+        {:command-types [] :remaining-sql nil}))))
+
+(defn- cmd-type-ddl? [cmd-type]
+  ;; Command types are organized with all DDL commands listed first, so all ddl commands are before ALTER_SEQUENCE.
+  ;; see https://github.com/h2database/h2database/blob/master/h2/src/main/org/h2/command/CommandInterface.java#L297
+  (< cmd-type CommandInterface/ALTER_SEQUENCE))
+
+(defn- contains-ddl? [{:keys [command-types remaining-sql]}]
+  (let [cmd-type-nums command-types]
+    (boolean
+     (or (some cmd-type-ddl? cmd-type-nums)
+         (some? remaining-sql)))))
+
+;; TODO: black-list RUNSCRIPT, and a bunch more -- but they're not technically ddl. Should be simple to build off of [[classify-query]].
+;; e.g.: similar to contains-ddl? but instead of cmd-type-ddl? use: #(#{CommandInterface/RUNSCRIPT} %)
+
+(defn- check-disallow-ddl-commands [{:keys [database] {:keys [query]} :native :as in}]
+  (let [query-classification (classify-query database query)]
+    (when (and query (contains-ddl? query-classification))
```

Comment:
`check-disallow-ddl-commands` wouldn't throw an error if `query` is nil before. Now it does, since `(classify-query nil)` raises an exception. 

So, either confirm `query` can never be nil when this function is called and replace the `query` check
```suggestion
    (when (contains-ddl? query-classification)
```
Or wrap the whole block in a `query` check, to preserve behaviour when `query` is nil
```
    (when query ... )
```

Reply:
Nice catch, thanks. Done